runtime/format.h

```diff
@@ -0,0 +1,108 @@
+//===-- runtime/format.h ----------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// FORMAT string processing
+
+#ifndef FORTRAN_RUNTIME_FORMAT_H_
+#define FORTRAN_RUNTIME_FORMAT_H_
+
+#include "terminator.h"
+#include "../lib/common/Fortran.h"
+#include <cinttypes>
+#include <optional>
+
+namespace Fortran::runtime {
+
+enum EditingFlags {
+  blankZero = 1,  // BLANK=ZERO or BZ edit
+  decimalComma = 2,  // DECIMAL=COMMA or DC edit
+  signPlus = 4,  // SIGN=PLUS or SP edit
+};
+
+struct MutableModes {
+  std::uint8_t editingFlags{0};  // BN, DP, SS
+  common::RoundingMode roundingMode{common::RoundingMode::TiesToEven};  // RN
+};
+
+// A single edit descriptor extracted from a FORMAT
+struct DataEdit {
+  char descriptor;  // capitalized: one of A, I, B, O, Z, F, E(N/S/X), D, G
+  char variation{'\0'};  // N, S, or X for EN, ES, EX
+  int width;  // the 'w' field
+  std::optional<int> digits;  // the 'm' or 'd' field
+  std::optional<int> expoDigits;  // 'Ee' field
+  MutableModes modes;
```

Comment:
While I agree keeping `MutableModes` inside `DataEdit`  makes passing data structure easier, I think it gives a "live" aspect to `DataEdit` that would make storing `DataEdit` in an attempt to prepocess formats weird. I would keep them separate.
That being said, I think storing `DataEdit` does not really make sens with the current approach because bigger changes would be needed to expose the iterations to the user, so that may just be fine to keep it there if you think it is better.

Reply:
These modes are the ones that can be changed within a FORMAT string, so either each data edit descriptor has to hold a copy of the current modes, or the FORMAT processor will have to interact with the context to ensure that the context always has the right modes.  Since the representation is small, associating the modes with the data edit descriptor makes the most sense to me.